docs/enhancements/agent-upgrade.md

```diff
@@ -0,0 +1,256 @@
+reference the tag where required, in particular for generation of metrics.
+
+In order to support both old and new agents the validation will not be a simple
+string comparison. Instead it will check if the `discovery_agent_version` sent
+by the agent is a full image reference or only a tag. If it is a full image
+reference then it will be a exact string comparision with the value of the
+`AGENT_DOCKER_IMAGE` environment variable. If it is only a tag, then it will be
+a comparison with the tag extracted from the value of the `AGENT_DOCKER_IMAGE`
+environment variable. In this last case a warning will be written to the log
+explaining that the comparision may be ureliable because only the tags were
+compared.
+
+To make this comparison useful it is important that the `AGENT_DOCKER_IMAGE`
+environment variable passed to the service doesn't use the `:latest` tag, or in
+general any moving tag, as that renders the comparison useless. In the _SaaS_
+stage environment that variable is already using a tag, so that shouldn't be a
+problem:
+
+```
+$ oc get deployment assisted-service -n assisted-installer-stage -o json | \
+jq '.spec.template.spec.containers[].env[] | select(.name == "AGENT_DOCKER_IMAGE")'
+{
+  "name": "AGENT_DOCKER_IMAGE",
+  "value": "registry-proxy.engineering.redhat.com/rh-osbs/openshift4-assisted-installer-agent-rhel8:v1.0.0-141"
+}
+```
+
+In the integration and development environments we are using the `:latest` tag,
+which effectively disables this feature.
+
+The service will check during startup the value of the `AGENT_DOCKER_IMAGE` and
+it will write a warning to the log if it is using the `:latest` tag.
+
+The description of the `compatible-agent` validation will be the following:
+
+> The installation cannot start just yet because this host's agent is in the
+> process of being upgraded to a newer version, please wait, this could take
+> a few minutes.
+
+A new `upgrade-agent` step will be added to the service and to the agent.
+
+When the service receives a request for next steps from a host it will check if
+that host is in one of the following states:
+
+- `added-to-existing-cluster`
+- `binding`
+- `disabled-unbound`
+- `disabled`
+- `disconnected-unbound`
+- `disconnected`
+- `discovering-unbound`
+- `discovering`
+- `installed`
+- `installing-in-progress`
+- `insufficient-unbound`
+- `insufficient`
+- `known-unbound`
+- `known`
+- `pending-for-input`
+- `preparing-failed`
+- `preparing-successful`
+- `resetting-pending-user-action`
+- `resetting`
+- `unbinding-pending-user-action`
+- `unbinding`
+
+If the host is in one of the above states then the service will check the result
+of the `compatible-agent` validation. If the validation passed then the service
+will return the usal next steps. If the validaton didn't pass then the service
+will return only the `upgrade-agent` step, replacing all the usual steps.
+
+If on the other hand the host is in one of the followin states:
+
+- `preparing-for-installation`
```

Comment:
But the host might have already got the install command (the host moves to installing status when the installer start running on the node)
Since the `preparing-for-installation` should be fairly short (few seconds) I think it's safer not to upgrade once the user clicked install.